File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/a-compiler-from-scratch-84427e53fe/job_byox_build_v1_f3e0d99cf97b485202190e3be4a0e8f6/attempt-001/sealed/reference/lib/pebble/parser.rb
# frozen_string_literal: true

module Pebble
  class Parser
    def initialize(tokens)
      @tokens = tokens
      @current = 0
    end

    def parse
      statements = []
      statements << statement until check(:EOF)
      consume(:EOF, "expected end of input")
      fail_at(peek, "unexpected token after EOF") unless @current == @tokens.length
      { type: :program, statements: statements }
    end

    private

    def statement
      return let_statement if match(:LET)
      return print_statement if match(:PRINT)
      return if_statement if match(:IF)
      return while_statement if match(:WHILE)
      return assignment_statement if check(:IDENTIFIER) && check_next(:EQUAL)

      fail_at(peek, "expected a statement")
    end

    def let_statement
      name = consume(:IDENTIFIER, "expected a variable name after 'let'")
      consume(:EQUAL, "expected '=' after variable name")
      value = expression
      consume(:SEMICOLON, "expected ';' after declaration")
      { type: :let, name: name.lexeme, value: value, token: name }
    end

    def assignment_statement
      name = consume(:IDENTIFIER, "expected a variable name")
      consume(:EQUAL, "expected '=' after variable name")
      value = expression
      consume(:SEMICOLON, "expected ';' after assignment")
      { type: :assign, name: name.lexeme, value: value, token: name }
    end

    def print_statement
      token = previous
      value = expression
      consume(:SEMICOLON, "expected ';' after printed expression")
      { type: :print, value: value, token: token }
    end

    def if_statement
      token = previous
      consume(:LEFT_PAREN, "expected '(' after 'if'")
      condition = expression
      consume(:RIGHT_PAREN, "expected ')' after condition")
      then_body = block
      else_body = match(:ELSE) ? block : []
      {
        type: :if,
        condition: condition,
        then_body: then_body,
        else_body: else_body,
        token: token
      }
    end

    def while_statement
      token = previous
      consume(:LEFT_PAREN, "expected '(' after 'while'")
      condition = expression
      consume(:RIGHT_PAREN, "expected ')' after condition")
      { type: :while, condition: condition, body: block, token: token }
    end

    def block
      consume(:LEFT_BRACE, "expected '{' before block")
      statements = []
      until check(:RIGHT_BRACE)
        fail_at(peek, "expected '}' after block") if check(:EOF)
        statements << statement
      end
      consume(:RIGHT_BRACE, "expected '}' after block")
      statements
    end

    def expression
      equality
    end

    def equality
      fold_binary(:comparison, :EQUAL_EQUAL, :BANG_EQUAL)
    end

    def comparison
      fold_binary(:term, :LESS, :LESS_EQUAL, :GREATER, :GREATER_EQUAL)
    end

    def term
      fold_binary(:factor, :PLUS, :MINUS)
    end

    def factor
      fold_binary(:unary, :STAR, :SLASH, :PERCENT)
    end

    def fold_binary(child_method, *operators)
      node = send(child_method)
      while match(*operators)
        operator = previous
        right = send(child_method)
        node = {
          type: :binary,
          operator: operator.type,
          left: node,
          right: right,
          token: operator
        }
      end
      node
    end

    def unary
      if match(:BANG, :MINUS)
        operator = previous
        return {
          type: :unary,
          operator: operator.type,
          operand: unary,
          token: operator
        }
      end
      primary
    end

    def primary
      if match(:INTEGER, :TRUE, :FALSE)
        token = previous
        return { type: :literal, value: token.literal, token: token }
      end
      if match(:IDENTIFIER)
        token = previous
        return { type: :variable, name: token.lexeme, token: token }
      end
      if match(:LEFT_PAREN)
        value = expression
        consume(:RIGHT_PAREN, "expected ')' after expression")
        return value
      end
      fail_at(peek, "expected an expression")
    end

    def match(*types)
      types.each do |type|
        next unless check(type)

        advance
        return true
      end
      false
    end

    def consume(type, message)
      return advance if check(type)

      fail_at(peek, message)
    end

    def check(type)
      token = peek
      token && token.type == type
    end

    def check_next(type)
      token = @tokens[@current + 1]
      token && token.type == type
    end

    def advance
      token = peek
      @current += 1 if token
      token
    end

    def peek
      @tokens[@current]
    end

    def previous
      @tokens[@current - 1]
    end

    def fail_at(token, message)
      if token && token.respond_to?(:line) && token.respond_to?(:column)
        raise ParseError.new(message, token.line, token.column)
      end
      raise ParseError.new(message, 1, 1)
    end
  end
end

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/from-source-code-to-machine-code-build-your-own-compiler-from-scratch-2877aeded6/job_byox_build_v1_7a07ec294de1cc10739ceacd586b1ffd/attempt-001/sealed/reference/minnow/parser.py
"""Recursive-descent parser with explicit precedence levels."""

from .errors import ParseError
from .model import Assign, Binary, Block, If, Let, Literal, Print, Program, TokenKind, Unary, Variable, While


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.current = 0

    def parse_program(self):
        statements = []
        while not self._check(TokenKind.EOF):
            statements.append(self._statement())
        self._consume(TokenKind.EOF, "expected end of input")
        return Program(tuple(statements))

    def _statement(self):
        if self._match(TokenKind.LET):
            return self._let_statement()
        if self._match(TokenKind.PRINT):
            return self._print_statement()
        if self._match(TokenKind.IF):
            return self._if_statement()
        if self._match(TokenKind.WHILE):
            return self._while_statement()
        if self._check(TokenKind.IDENT) and self._check_next(TokenKind.EQUAL):
            return self._assign_statement()
        self._error(self._peek(), "expected statement")

    def _let_statement(self):
        name = self._consume(TokenKind.IDENT, "expected name after 'let'")
        self._consume(TokenKind.EQUAL, "expected '=' after declaration name")
        initializer = self._expression()
        self._consume(TokenKind.SEMICOLON, "expected ';' after declaration")
        return Let(name, initializer)

    def _assign_statement(self):
        name = self._consume(TokenKind.IDENT, "expected assignment name")
        self._consume(TokenKind.EQUAL, "expected '=' after assignment name")
        value = self._expression()
        self._consume(TokenKind.SEMICOLON, "expected ';' after assignment")
        return Assign(name, value)

    def _print_statement(self):
        value = self._expression()
        self._consume(TokenKind.SEMICOLON, "expected ';' after value")
        return Print(value)

    def _if_statement(self):
        self._consume(TokenKind.LEFT_PAREN, "expected '(' after 'if'")
        condition = self._expression()
        self._consume(TokenKind.RIGHT_PAREN, "expected ')' after condition")
        then_branch = self._block()
        else_branch = self._block() if self._match(TokenKind.ELSE) else None
        return If(condition, then_branch, else_branch)

    def _while_statement(self):
        self._consume(TokenKind.LEFT_PAREN, "expected '(' after 'while'")
        condition = self._expression()
        self._consume(TokenKind.RIGHT_PAREN, "expected ')' after condition")
        return While(condition, self._block())

    def _block(self):
        self._consume(TokenKind.LEFT_BRACE, "expected '{' before block")
        statements = []
        while not self._check(TokenKind.RIGHT_BRACE) and not self._check(TokenKind.EOF):
            statements.append(self._statement())
        self._consume(TokenKind.RIGHT_BRACE, "expected '}' after block")
        return Block(tuple(statements))

    def _expression(self):
        return self._equality()

    def _equality(self):
        expression = self._comparison()
        while self._match(TokenKind.EQUAL_EQUAL, TokenKind.BANG_EQUAL):
            operator = self._previous()
            expression = Binary(expression, operator, self._comparison())
        return expression

    def _comparison(self):
        expression = self._term()
        while self._match(TokenKind.LESS, TokenKind.LESS_EQUAL, TokenKind.GREATER, TokenKind.GREATER_EQUAL):
            operator = self._previous()
            expression = Binary(expression, operator, self._term())
        return expression

    def _term(self):
        expression = self._factor()
        while self._match(TokenKind.PLUS, TokenKind.MINUS):
            operator = self._previous()
            expression = Binary(expression, operator, self._factor())
        return expression

    def _factor(self):
        expression = self._unary()
        while self._match(TokenKind.STAR, TokenKind.SLASH, TokenKind.PERCENT):
            operator = self._previous()
            expression = Binary(expression, operator, self._unary())
        return expression

    def _unary(self):
        if self._match(TokenKind.BANG, TokenKind.MINUS):
            return Unary(self._previous(), self._unary())
        return self._primary()

    def _primary(self):
        if self._match(TokenKind.INT):
            token = self._previous()
            return Literal(token.value, token)
        if self._match(TokenKind.TRUE):
            token = self._previous()
            return Literal(1, token)
        if self._match(TokenKind.FALSE):
            token = self._previous()
            return Literal(0, token)
        if self._match(TokenKind.IDENT):
            return Variable(self._previous())
        if self._match(TokenKind.LEFT_PAREN):
            expression = self._expression()
            self._consume(TokenKind.RIGHT_PAREN, "expected ')' after expression")
            return expression
        self._error(self._peek(), "expected expression")

    def _match(self, *kinds):
        for kind in kinds:
            if self._check(kind):
                self._advance()
                return True
        return False

    def _consume(self, kind, message):
        if self._check(kind):
            return self._advance()
        self._error(self._peek(), message)

    def _check(self, kind):
        return self._peek().kind is kind

    def _check_next(self, kind):
        if self.current + 1 >= len(self.tokens):
            return False
        return self.tokens[self.current + 1].kind is kind

    def _advance(self):
        token = self._peek()
        if self.current < len(self.tokens):
            self.current += 1
        return token

    def _peek(self):
        return self.tokens[self.current]

    def _previous(self):
        return self.tokens[self.current - 1]

    @staticmethod
    def _error(token, message):
        raise ParseError(message, line=token.line, column=token.column)


def parse(tokens):
    return Parser(tokens).parse_program()
